src/libraries/System.IO.FileSystem/tests/FileStream/FileStreamConformanceTests.Windows.cs

```diff
@@ -176,4 +176,152 @@ public struct SHARE_INFO_502
+                    // which is fine, we just want to make sure that reading from devices is supported (#54143)
+                }
+            }
+        }
+
+        private static FileStream? OpenFirstAvailableDeviceInterface()
+        {
+            const int DIGCF_PRESENT = 0x2;
+            const int DIGCF_DEVICEINTERFACE = 0x10;
+            const int ERROR_NO_MORE_ITEMS = 259;
+
+            HidD_GetHidGuid(out Guid HidGuid);
+            IntPtr deviceInfoSet = SetupDiGetClassDevs(in HidGuid, IntPtr.Zero, IntPtr.Zero, DIGCF_PRESENT | DIGCF_DEVICEINTERFACE);
+
+            try
+            {
+                SP_DEVINFO_DATA deviceInfoData = new SP_DEVINFO_DATA();
+                deviceInfoData.cbSize = (uint)Marshal.SizeOf(deviceInfoData);
+
+                uint deviceIndex = 0;
+                while (SetupDiEnumDeviceInfo(deviceInfoSet, deviceIndex++, ref deviceInfoData))
+                {
+                    if (Marshal.GetLastWin32Error() == ERROR_NO_MORE_ITEMS)
+                    {
+                        break;
+                    }
+
+                    SP_DEVICE_INTERFACE_DATA deviceInterfaceData = new SP_DEVICE_INTERFACE_DATA();
+                    deviceInterfaceData.cbSize = Marshal.SizeOf(deviceInterfaceData);
+
+                    if (!SetupDiEnumDeviceInterfaces(deviceInfoSet, IntPtr.Zero, in HidGuid, deviceIndex, ref deviceInterfaceData))
+                    {
+                        continue;
+                    }
+
+                    SP_DEVICE_INTERFACE_DETAIL_DATA deviceInterfaceDetailData = new SP_DEVICE_INTERFACE_DETAIL_DATA();
+                    deviceInterfaceDetailData.cbSize = IntPtr.Size == 8 ? 8 : 6;
+
+                    uint size = (uint)Marshal.SizeOf(deviceInterfaceDetailData);
+
+                    if (!SetupDiGetDeviceInterfaceDetail(deviceInfoSet, ref deviceInterfaceData, ref deviceInterfaceDetailData, size, ref size, IntPtr.Zero))
+                    {
+                        continue;
+                    }
+
+                    string devicePath = deviceInterfaceDetailData.DevicePath;
+                    Assert.StartsWith(@"\\?\hid", devicePath);
+
+                    try
+                    {
+                        return new FileStream(devicePath, FileMode.Open, FileAccess.Read, FileShare.Read, 0, FileOptions.Asynchronous);
+                    }
+                    catch (IOException)
+                    {
+                        // device has been locked by another process
+                        continue;
+                    }
+                }
+            }
+            finally
+            {
+                SetupDiDestroyDeviceInfoList(deviceInfoSet);
+            }
+
+            return null;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        struct SP_DEVICE_INTERFACE_DATA
+        {
+            public int cbSize;
+            public Guid interfaceClassGuid;
+            public int flags;
+            private nuint reserved;
```

Comment:
do we have guidelines somewhere about when to use nint/nuint vs IntPtr/UIntPtr..

Reply:
> do we have guidelines somewhere about when to use nint/nuint vs IntPtr/UIntPtr..

That is a very good question, but I don't know the answer. Perhaps @stephentoub  or @jkotas know?

Reply:
I do not think we have it written anywhere. I think you pretty much never want to use IntPtr/UIntPtr in new code. If the value is size or offset, use nint/uint. If the value is pointer, use actual pointer type.

The few situations where it may be still appropriate to use IntPtr/UIntPtr is old code and APIs where actual pointer type would introduce a lot of unnecessary casts; or where IntPtr/UIntPtr is used as a handle (these two cases have large overlap).

Reply:
There's this: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nint-nuint